backend/src/main/java/ai/verta/modeldb/utils/RdbmsUtils.java

```diff
@@ -237,13 +237,59 @@ public static ArtifactEntity generateArtifactEntity(
   }
 
   public static ObservationEntity generateObservationEntity(
-      Object entity, String fieldType, Observation observation)
+      Session session,
+      Object entity,
+      String fieldType,
+      Observation observation,
+      String entity_name,
+      String entity_id)
       throws InvalidProtocolBufferException {
-    return new ObservationEntity(entity, fieldType, observation);
+    if (session == null
+        || observation.hasEpochNumber()) { // request came from creating the entire ExperimentRun
+      return new ObservationEntity(entity, fieldType, observation);
+    } else {
+      if (entity_name.equalsIgnoreCase("ExperimentRunEntity") && observation.hasAttribute()) {
+        String MAX_EPOCH_NUMBER_SQL =
```

Comment:
I hope this converts to a join.

Reply:
it should, will verify, though what is the other way to process the query if not join?

Reply:
verified it uses existing index and converts to hash join on large data

Reply:
An inefficient join could have happened. Thanks for checking